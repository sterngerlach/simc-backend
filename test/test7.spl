
{
    type Vector3 = int[3];
    Vector3 x;
    Vector3 y;
    Vector3 z;
    
    void setVector(Vector3 v, int x, int y, int z)
    {
        v[0] = x;
        v[1] = y;
        v[2] = z;
    }

    void addVector(Vector3 z, Vector3 x, Vector3 y)
    {
        int i;

        for (i = 0 .. 3)
            z[i] = x[i] + y[i];
    }

    int dotProduct(Vector3 x, Vector3 y)
    {
        int result = 0;
        int i;

        for (i = 0 .. 3)
            result += x[i] * y[i];

        return result;
    }

    int squaredDistance(Vector3 x, Vector3 y)
    {
        int result = 0;
        int i;

        for (i = 0 .. 3)
            result += (x[i] - y[i]) * (x[i] - y[i]);

        return result;
    }

    void printVector(Vector3 v)
    {
        int i;

        sprint("[");

        for (i = 0 .. 3) {
            iprint(v[i]);
            sprint(", ");
        }

        sprint("\b\b]\n");
    }
    
    new(x);
    new(y);
    new(z);
    
    setVector(x, 2, 3, 4);
    setVector(y, 4, 8, 12);
    
    sprint("x: ");
    printVector(x);
    sprint("y: ");
    printVector(y);
    
    addVector(z, x, y);
    sprint("x + y: ");
    printVector(z);

    sprint("x * y: ");
    iprint(dotProduct(x, y));
    sprint("\n");

    sprint("|x - y| ^ 2 = ");
    iprint(squaredDistance(x, y));
    sprint("\n");
}

