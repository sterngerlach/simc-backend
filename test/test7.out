x: [2, 3, 4, ]
y: [4, 8, 12, ]
x + y: [6, 11, 16, ]
x * y: 80
|x - y| ^ 2 = 93
